Suggested updates to contribution guide (#214)
viash-hub/biobox: docs/PYTHON_SCRIPT_DEVELOPMENT.md

# Python Script Development Guide

> This guide covers best practices for writing Python-based component scripts in biobox. Most biobox components are bash wrappers (see [Script Development Guide](SCRIPT_DEVELOPMENT.md)); Python is used when the wrapper needs real logic — parsing structured output, conditionals beyond a few branches, or data transformations that would be painful in bash.

Reviewer: ```suggestion
This guide covers best practices for writing Python-based component scripts in biobox. Most biobox components are bash wrappers (see [Script Development Guide](SCRIPT_DEVELOPMENT.md)); Python is used when the wrapper needs real logic and/or complex data structures — parsing structured output, conditionals beyond a few branches, or data transformations that would be painful in bash.
```

## Table of Contents
- [When to Choose Python over Bash](#when-to-choose-python-over-bash)
- [Script Structure and Template](#script-structure-and-template)
- [The `par` and `meta` Dicts](#the-par-and-meta-dicts)
- [Argument Handling](#argument-handling)
- [Calling External Tools](#calling-external-tools)
- [Resource Usage](#resource-usage)
- [Error Handling and Logging](#error-handling-and-logging)
- [File and Path Handling](#file-and-path-handling)
- [Dependencies and Containers](#dependencies-and-containers)
- [Testing](#testing)
- [Real-World Example](#real-world-example)
- [Common Pitfalls](#common-pitfalls)

## When to Choose Python over Bash

Default to bash. Reach for Python when:

> - The wrapper parses structured output (JSON, YAML, TSV with complex rules) and feeds it back to the tool.

Reviewer: ```suggestion
- The wrapper parses structured data (JSON, YAML, TSV with complex rules) and feeds it to the tool.
```

> - Arg construction has non-trivial branching (more than a handful of if/else).

Reviewer: ```suggestion
- Argument construction has non-trivial branching (more than a handful of if/else).
```

- The wrapper has to coordinate multiple tool invocations and shuffle their outputs.
- Input validation is substantive (e.g. matched paired-end file lists, schema checks).

> If the job is "arrange CLI flags and exec one tool", stay in bash. It is shorter and the surrounding ecosystem in this repo assumes bash.

Reviewer: ```suggestion
If the job is "arrange CLI flags and execute one tool", stay in bash. It is shorter and the surrounding ecosystem in this repo assumes bash.
```

In the config, use a Python resource:

```yaml
resources:
  - type: python_script
    path: script.py
```

## Script Structure and Template

```python
import subprocess
from pathlib import Path

## VIASH START
par = {
    "input": "test.fastq",
    "output": "out.bam",
    "verbose": False,
}
meta = {
    "cpus": 4,
    "memory_gb": 8,
    "temp_dir": "/tmp",
    "resources_dir": ".",
}
## VIASH END

# ... your logic ...

cmd = [
    "your_tool",
    "--input", par["input"],
    "--output", par["output"],
]
if par.get("verbose"):
    cmd.append("--verbose")
if meta.get("cpus"):
    cmd += ["--threads", str(meta["cpus"])]

> subprocess.run(cmd, check=True)

Reviewer: This will not show any output from the program (i.e. when there is output written to stderr)

For cellranger multi, I did the following:

```python
with (
    subprocess.Popen(
        cmd,
        text=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        bufsize=1,
        cwd=temp_dir,
    ) as process,
    (par["output"] / "cellranger_multi.log").open("w") as open_log,
):
    for line in process.stdout:
        print(line.strip(), flush=True)
        open_log.write(line)
```
This writes both the output to a log file _and_ to the stdout of the component _on the fly_ (e.g. while the executble has not finished yet, nice for longer running processes)
But it depends on what the executable with its stdout:

- executable outputs data to stdout
- executable outputs logging to stdout/stderr and writes its own output files

Edit: I saw that this was done below. Still might be handy to cover this somewhere

Author: Could you please add this to the doc?

- `par["<arg_name>"]` — the value of an argument (without the leading `--`, and with dashes replaced by underscores).
- Missing-or-not-provided optional arguments are `None`, **not** absent from the dict.
- Booleans arrive as Python `bool` values (`True` / `False`), not the strings `"true"` / `"false"`.
- `multiple: true` arguments arrive as a Python `list` of the appropriate type.
- File arguments arrive as `str` paths; wrap in `pathlib.Path` if you need path operations.

`meta` mirrors the bash `meta_*` variables. The common keys:

| Key | Type | Purpose |
| --- | --- | --- |
| `meta["cpus"]` | `int` or `None` | CPU core budget |
| `meta["memory_gb"]`, `meta["memory_mb"]`, ... | `int` or `None` | Memory budgets in various units |
| `meta["temp_dir"]` | `str` | Scratch directory |
| `meta["resources_dir"]` | `str` | Path to component resource files |
| `meta["name"]` | `str` | Component name |
| `meta["config"]` | `str` | Path to the compiled config (useful for introspection) |

## Argument Handling

```python
# Required argument — Viash guarantees this is not None
cmd = ["tool", "--input", par["input"]]

# Optional argument

> if par.get("reference") is not None:

Reviewer: The keys in `par` are always there I think.
```
if par["reference"] is not None:
```
And almost always it is net to use
```
if par["reference"]:
```

This has the advantage that:

- Booleans are already covered
- Empty lists are also evaluated correctly

However, there is one exception for integers:

`if 0` evaluates to False, so in this case one might need to to `if par["arg"] or par["arg"] == 0`, depending

Author: Could you please add this to the doc yourself?

## Calling External Tools

Always `subprocess.run(cmd, check=True)` with `cmd` as a list:

```python
subprocess.run(
    ["samtools", "view", "-b", par["input"], "-o", par["output"]],
    check=True,
)
```

- Never `shell=True` unless you have a very good reason (piping through shell operators, and even then prefer composing in Python).
- `check=True` turns non-zero exits into `CalledProcessError`, so failures stop the script.
- Capture output with `capture_output=True, text=True` when you need to parse it; otherwise let it stream to the component's stderr/stdout.

For commands that need the shell (pipelines, redirections), prefer Python-level plumbing:

```python
# Instead of: subprocess.run("tool | grep -v '^#' > out.tsv", shell=True, check=True)
with open(par["output"], "w") as out:
    p1 = subprocess.Popen(["tool"], stdout=subprocess.PIPE)
    subprocess.run(["grep", "-v", "^#"], stdin=p1.stdout, stdout=out, check=True)
    p1.stdout.close()
    p1.wait()
```

## Resource Usage

```python
cmd = ["tool", "--input", par["input"], "--output", par["output"]]

if meta.get("cpus"):
    cmd += ["--threads", str(meta["cpus"])]
if meta.get("memory_gb"):

>     cmd += ["--memory", f"{meta['memory_gb']}G"]

Reviewer: ```suggestion
if meta["cpus"]:
    cmd += ["--threads", str(meta["cpus"])]
if meta["memory_gb"]:
    cmd += ["--memory", f"{meta['memory_gb']}G"]
```
